File: array/35.c
/*
 * 典型binary search O(logn)
 */
int binarySearch(int* nums, int target, int start, int end)
{
    int mid = (start + end) / 2;
    if(end < start) return start;
    if(target == nums[mid]) return mid;
    else if(target < nums[mid]) return binarySearch(nums, target, start, mid - 1);
    else return binarySearch(nums, target, mid + 1, end);
    
}
int searchInsert(int* nums, int numsSize, int target) {
    return binarySearch(nums, target, 0, numsSize - 1);
}

File: array/35.go
func searchInsert(nums []int, target int) int {
    if target < nums[0] {
        return 0
    }
    if target > nums[len(nums)-1] {
        return len(nums)
    }
    var start, end, mid, index int
    start = 0   
    end = len(nums) - 1
    for start != end {
        mid = start + (end - start) / 2
        if target == nums[mid] {
            return mid
        }else if target > nums[mid] {
            start = mid + 1
            index = start
        }else {
            end = mid
            index = end
        }
    }
    return index
}
